### PR TITLE
Flink job migration 1.1 -> 1.2 documentation

### PR DESCRIPTION
**Pull request by StefanRRichter**
Documentation how to bring a job from Flink 1.1 to Flink 1.2

### CODE REVIEW

File: docs/ops/upgrading.md [hunk 1]
```diff
@@ -35,23 +35,105 @@ under the License.
...
+This section describes the general way of upgrading Flink framework version from version 1.1.x to 1.2.x and migrating your
+jobs between the two versions.
+
+In a nutshell, this procedure consists of 2 fundamental steps:
+
+1. Take a savepoint in Flink 1.1.x for the jobs you want to migrate.
+2. Resume your jobs under Flink 1.2.x from the previously taken savepoints.
+
+Besides those two fundamental steps, some additional steps can be required that depend on the way you want to change the
+Flink version. In this guide we differentiate two approaches to upgrade from Flink 1.1.x to 1.2.x: **inplace** upgrade and
```
Comment (alpinegizmo) [inline docs/ops/upgrading.md]: **in-place** would be better than **inplace** (here and also in the next paragraph)

File: docs/ops/upgrading.md [hunk 2]
```diff
@@ -35,23 +35,105 @@ under the License.
...
+  2. Resume from the savepoints with the new Flink 1.2.x installation.
+  3. If everything runs ok, stop and shutdown the old Flink 1.1.x cluster.
+
+In the following, we will first present the preconditions for successful job migration and then go into more detail 
+about the steps that we outlined before.
+
+### Preconditions
+
+Before starting the migration, please check that the jobs you are trying to migrate are following the
+best-practises for [savepoints]({{ site.baseurl }}/setup/savepoints.html). In particular, we advise you to check that
```
Comment (alpinegizmo) [inline docs/ops/upgrading.md]: best practices

rather than best-practises

File: docs/ops/upgrading.md [hunk 3]
```diff
@@ -35,23 +35,105 @@ under the License.
...
+This is a *soft* precondition, and restore *should* still work in case you forgot about assigning `uid`s. 
+If you run into a case where this is not working, you can *manually* add the generated legacy vertex ids from Flink 1.1 
+to your job using the `setUidHash(String hash)` call. For each operator (in operator chains: only head operator) you 
+must assign the 32 character hex string representing the hash that you can see in the web ui or logs for the operator.
+
+Besides operator uids, there are currently three *hard* preconditions for job migration that will make migration fail: 
+
+1. as mentioned in earlier release notes, we do not support migration for state in RocksDB that was checkpointed using 
+`semi-asynchronous` mode. In case your old job was using this mode, you can still change your job to use 
+`fully-asynchronous` mode before taking the savepoint that is used as basis for the migration.
```
Comment (alpinegizmo) [inline docs/ops/upgrading.md]: used as the basis

File: docs/ops/upgrading.md [hunk 4]
```diff
@@ -35,23 +35,105 @@ under the License.
...
+
+2. The CEP operator is currently not supported for migration. If your job uses this operator you can (curently) not 
+migrate it. We are planning to provide migration support for CEP operator in a later bugfix release.
```
Comment (alpinegizmo) [inline docs/ops/upgrading.md]: for the CEP operator

File: docs/ops/upgrading.md [hunk 5]
```diff
@@ -35,23 +35,105 @@ under the License.
...
+
+### Preconditions
+
+Before starting the migration, please check that the jobs you are trying to migrate are following the
+best-practises for [savepoints]({{ site.baseurl }}/setup/savepoints.html). In particular, we advise you to check that 
+explicit `uid`s were set for operators in your job. 
+
+This is a *soft* precondition, and restore *should* still work in case you forgot about assigning `uid`s. 
+If you run into a case where this is not working, you can *manually* add the generated legacy vertex ids from Flink 1.1 
+to your job using the `setUidHash(String hash)` call. For each operator (in operator chains: only head operator) you
```
Comment (alpinegizmo) [inline docs/ops/upgrading.md]: only the head operator

File: docs/ops/upgrading.md [hunk 6]
```diff
@@ -35,23 +35,105 @@ under the License.
...
+
+1. as mentioned in earlier release notes, we do not support migration for state in RocksDB that was checkpointed using 
+`semi-asynchronous` mode. In case your old job was using this mode, you can still change your job to use 
+`fully-asynchronous` mode before taking the savepoint that is used as basis for the migration.
+
+2. The CEP operator is currently not supported for migration. If your job uses this operator you can (curently) not 
+migrate it. We are planning to provide migration support for CEP operator in a later bugfix release.
+
+3. Another **important** precondition is that all the savepoint data is accessible from the new installation and resides under
+the same absolute path. Please notice that the savepoint data is typically not self contained in just the created
```
Comment (alpinegizmo) [inline docs/ops/upgrading.md]: Another **important** precondition is that all the savepoint data must be accessible from the new installation and reside under the same absolute path.

File: docs/ops/upgrading.md [hunk 7]
```diff
@@ -35,23 +35,105 @@ under the License.
...
+
+3. Another **important** precondition is that all the savepoint data is accessible from the new installation and resides under
+the same absolute path. Please notice that the savepoint data is typically not self contained in just the created 
+savepoint file. Additional files can be referenced from inside the savepoint file (e.g. the output from state backend 
+snapshots)! There is currently no simple way to identify and move all data that belongs to a savepoint.
+
+
+### STEP 1: Taking a savepoint in Flink 1.1.x.
+
+First major step in job migration is taking a savepoint of your jon running in Flink 1.1.x. You can do this with the
```
Comment (alpinegizmo) [inline docs/ops/upgrading.md]: jon -> job

File: docs/ops/upgrading.md [hunk 8]
```diff
@@ -35,23 +35,105 @@ under the License.
...
+
+In this step, we update the framework version of the cluster. What this basically means is replacing the content of
+the Flink installation with the new version. This step can depend on how you are running Flink in your cluster (e.g. 
+standalone, on Mesos, ...).
+
+If you are unfamiliar with installing Flink in your cluster, please read the [deployment and cluster setup documentation]({{ site.baseurl }}/setup/index.html).
+
+### STEP 3: Resuming the job under Flink 1.2.x from Flink 1.1.x savepoint.
+
+As last step of job migration, you resume the from the taken savepoint on a cluster that runs Flink 1.2.x. You can do
```
Comment (alpinegizmo) [inline docs/ops/upgrading.md]: As the last step of job migration, you resume from the savepoint taken above on the updated cluster.

File: docs/ops/upgrading.md [hunk 9]
```diff
@@ -35,23 +35,105 @@ under the License.
...
 
 
 
-## Special Considerations for Upgrades from Flink 1.1.x to Flink 1.2.x
-
-  - The parallelism of the Savepoint in Flink 1.1.x becomes the maximum parallelism in Flink 1.2.x.
-  - Increasing the parallelism for upgraded jobs is not possible out of the box.
+## Limitations and Special Considerations for Upgrades from Flink 1.1.x to Flink 1.2.x
+  
+  - The maximum parallelism of a job that was migrated from Flink 1.1.x to 1.2.x currently fixed as the parallelism of
```
Comment (alpinegizmo) [inline docs/ops/upgrading.md]: is currently fixed

File: docs/ops/upgrading.md [hunk 10]
```diff
@@ -35,23 +35,105 @@ under the License.
...
 
 
-## Special Considerations for Upgrades from Flink 1.1.x to Flink 1.2.x
-
-  - The parallelism of the Savepoint in Flink 1.1.x becomes the maximum parallelism in Flink 1.2.x.
-  - Increasing the parallelism for upgraded jobs is not possible out of the box.
+## Limitations and Special Considerations for Upgrades from Flink 1.1.x to Flink 1.2.x
+  
+  - The maximum parallelism of a job that was migrated from Flink 1.1.x to 1.2.x currently fixed as the parallelism of 
+  the job. This means that the parallelism can not be increased after migration. This limitation might be fixed in a
```
Comment (alpinegizmo) [inline docs/ops/upgrading.md]: This limitation might be removed in a future bugfix release.

**Comment by StefanRRichter** [comment]
Thanks for your feedback @alpinegizmo! I made another commit that contains the suggested corrections.

**Comment by uce** [comment]
Looks good to me. Going to merge this.

**Pull request closed without merging**